src/nvim/os/env.c

```diff
@@ -427,6 +427,31 @@ char *vim_getenv(const char *name)
     return xstrdup(kos_env_path);
   }
 
+  // XDG Base Directory expansion.
+  if (strncmp(name, "XDG_", 4) == 0) {
+    const char *home = vim_getenv("HOME");
+    if (!home) {
+        return NULL;
+    }
+
+    const char *home_suffix = "";
+    if (strcmp(name, "XDG_DATA_HOME") == 0) {
+      home_suffix = "/.local/share/";
+    } else if (strcmp(name, "XDG_CONFIG_HOME") == 0) {
+      home_suffix = "/.config/";
+    } else if (strcmp(name, "XDG_CACHE_HOME") == 0) {
+      home_suffix = "/.cache/";
+    } else if (strcmp(name, "XDG_RUNTIME_DIR") == 0) {
+      return xstrdup("/tmp/");  // Arbitrary value, any idea?
+    }
+
+    // We need to apprend the prefix to the HOME path.
+    char *value = vim_getenv("HOME");
+    value = xrealloc(value, strlen(value) + strlen(home_suffix) + 1);
+    strcat(value, home_suffix);
```

Comment:
If you already have to compute length better use

    const size_t value_len = strlen(value);
    const size_t home_suffix_len = strlen(home_suffix);
    value = xrealloc(value, value_len + home_suffix_len + 1);
    memmove(value + value_len, home_suffix, home_suffix_len + 1);